crates/adapters/src/format/avro/deserializer.rs

```diff
@@ -0,0 +1,1596 @@
+            Err(de::Error::custom("a record must have a least one field")),
+            |item| match (item.0.as_ref(), &item.1) {
+                ("type", Value::String(x)) | ("type", Value::Enum(_, x)) => Ok((
+                    seed.deserialize(StringDeserializer {
+                        input: x.to_owned(),
+                    })?,
+                    self,
+                )),
+                (field, Value::String(_)) => Err(de::Error::custom(format!(
+                    "expected first field named 'type': got '{field}' instead"
+                ))),
+                (_, _) => Err(de::Error::custom(
+                    "expected first field of type String or Enum for the type name".to_string(),
+                )),
+            },
+        )
+    }
+}
+
+fn unwrap_union(v: &Value) -> &Value {
+    match v {
+        Value::Union(_, v) => v.as_ref(),
+        _ => v,
+    }
+}
+
+impl<'de> de::VariantAccess<'de> for EnumDeserializer<'de> {
+    type Error = Error;
+
+    fn unit_variant(self) -> Result<(), Error> {
+        Ok(())
+    }
+
+    fn newtype_variant_seed<T>(self, seed: T) -> Result<T::Value, Error>
+    where
+        T: DeserializeSeed<'de>,
+    {
+        self.input.get(1).map_or(
+            Err(de::Error::custom(
+                "expected a newtype variant, got nothing instead.",
+            )),
+            |item| seed.deserialize(&Deserializer::new(&item.1)),
+        )
+    }
+
+    fn tuple_variant<V>(self, _len: usize, visitor: V) -> Result<V::Value, Error>
+    where
+        V: Visitor<'de>,
+    {
+        self.input.get(1).map_or(
+            Err(de::Error::custom(
+                "expected a tuple variant, got nothing instead.",
+            )),
+            |item| de::Deserializer::deserialize_seq(&Deserializer::new(&item.1), visitor),
+        )
+    }
+
+    fn struct_variant<V>(
+        self,
+        fields: &'static [&'static str],
+        visitor: V,
+    ) -> Result<V::Value, Error>
+    where
+        V: Visitor<'de>,
+    {
+        self.input.get(1).map_or(
+            Err(de::Error::custom("expected a struct variant, got nothing")),
+            |item| {
+                de::Deserializer::deserialize_struct(
+                    &Deserializer::new(&item.1),
+                    "",
+                    fields,
+                    visitor,
+                )
+            },
+        )
+    }
+}
+
+impl<'a, 'de> de::Deserializer<'de> for &'a Deserializer<'de> {
+    type Error = Error;
+
+    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        match unwrap_union(self.input) {
+            Value::Null => visitor.visit_unit(),
+            &Value::Boolean(b) => visitor.visit_bool(b),
+            Value::Int(i) | Value::Date(i) => visitor.visit_i32(*i),
+            Value::Long(i)
+            | Value::TimeMicros(i)
+            | Value::TimestampMicros(i)
+            //| Value::TimestampNanos(i)
```

Comment:
are these TODO?

Reply:
the crate I copied this from supports time in nanoseconds, which doesn't seem to be part of the Avro spec. Not sure what it's for.